### DISCUSSION TITLE
compositing many images - pixel cache?

### DISCUSSION THREAD

**Discussion by thebobcampbell** [ImageMagick/ImageMagick#4260] [Help]
TL;DR:
How to merge a large number of files with convert/compose/composite by making use of a pixel cache, or is there some other, better way than a for loop calling convert a thousand times like I'm doing now?

A while ago, I started making timelapse videos from still images.  Then I discovered someone who had made composite images from timelapses, and I decided to try myself.  I've written a script to sort and combine images into timelapse video, composites, and a video of the composite being created.
Mainly, I'm interested in the composite now.
The script I wrote is available here: (warning - it's a mess)
https://github.com/thebobcampbell/mklapse
I'm in the process of cleaning it up and rewriting it, so I began looking into more effective ways to build the composite.
Currently, the script goes through the stack of images, sorts by datestamp in the EXIF data (not by filename), and runs through a bash 'for' loop adding the next image into a placeholder image and outputting to a third tmp image, via:

```
  for image in `cat ${input}`
  do
    convert -compose ${method} -composite $image ${temp}.${extension} tmp.${extension} 2>/dev/null
[...]
          mv tmp.${extension} ${temp}.${extension}
  done
```

Unfortunately, that takes a long time, especially if your files are 50MB+ TIFFs.
So recently (after using IM since nearly the beginning), I've discovered the "pixel cache".  This seems to be intriguing for what I'm doing, but the only references I've found are for the API, or usages that just aren't as specific as this.

Mind you, I'm compositing anywhere from several hundred to a couple thousand images at once.  BUT the "temp" image I use as an interim would benefit greatly if it were just kept in memory and I could feed a thousand images into convert/magick.

I'm currently using:
Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules
Delegates (built-in): bzlib djvu fftw fontconfig freetype gslib heic jbig jng jp2 jpeg lcms ltdl lzma openexr png ps raw tiff webp x xml zlib
on MacOS Catalina.  I also have Linux Mint (Ubuntu variant) available as well, so I could do this in IM7 - answers for either are appreciated!

Sample images
![2014-03-28-composite-edit](https://user-images.githubusercontent.com/120732/134290781-7ab0f001-1ca2-4dec-8664-dbb6f9af36f7.jpg)
![2016-05-09](https://user-images.githubusercontent.com/120732/134290788-c45eb23a-fd66-43cf-90f2-081ff7192a4f.jpg)
![2019-09-21-Moab-stars-D850-take6](https://user-images.githubusercontent.com/120732/134290794-77b7fd4f-9e8f-48c8-b2c7-ef880df83bcb.jpg)
:

**Comment by snibgo** [comment]
If you use IM v7, I suggest using "magick" instead of "convert". The order is important: first read the inputs, then do "-compose ${method} -composite", then write the output.

You have a convert followed by a mv (to rename the output to one of the inputs). You don't need a mv. Instead, your convert can use the same filename for an input and the output.

Fastest performance will probably come from using "miff" or "mpc" as the extension.

You write and then immediately read a temporary file. Instead, you can do this (bash syntax):
```
magick
  in1.ext \
  in2.ext \
  -compose ${method} -composite \
  in3.ext \
  -compose ${method} -composite \
  in4.ext \
  -compose ${method} -composite \
:
:
  out.ext
```
This executes magick only once, and uses no temporary files.

Of course, you would have a main script with a loop through the input file names that creates that long command, and then executes it. It might create an environment variable with the command, or write it to a temporary file.

An alternative is to write a IM script, which is called from magick with "-script".

EDIT to add a point about pixel caches. These are structures that IM uses to store pixels in. These are like 2D arrays. When we read an image from a file, IM uncompresses the image and stores the pixel values in a pixel cache. Writing a file is the opposite. When we process an image, this may update a pixel cache in place, or may create a new pixel cache that is gradually populated from the old one. A pixel cache may be in memory or on disk. IM always uses pixel caches, whether we use the CLI or C interface. A pixel cache is not "remembered" from one command to another.

See https://www.imagemagick.org/script/architecture.php

**Reply by thebobcampbell** [reply]
Great, thanks! I'm sure I could create the command line within the bash script.  I haven't had a chance to look into IM7 scripting, so creating the command might be easier to implement.

I assume the magick command example you give is not possible with IM6 convert?

That's what I figured about the pixel cache (that it wouldn't be persistent across my executions of convert) so that confirms that I could probably see a great increase in performance using a single call to magick (maybe convert?).

Thanks again!

**Reply by thebobcampbell** [reply]
Tested creating the command with convert (same as above magick command, but with 'convert') and it worked!  And it was considerably faster.  Now I'll have to make sure I have my pixel cache settings optimized.

**Comment by snibgo** [comment]
Yes, my comments apply to IM v6 "convert" as well  as v7 "magick", except for "-script" which v6 doesn't have.

The only advantage to "-script" is that there is no limit to the length of the command.

The command I showed doesn't read files, or write files, or use more memory than the logical minimum. Of course there is an small overhead in creating the command, but that should be practically nothing.
